### ISSUE TITLE
Bug in recipes/quickstart/inference/local_inference/multi_modal_infer.py Results in "end_header_id|>" Preceding Generated Text

### ISSUE THREAD

**Issue by eii-lyl** [meta-llama/llama-recipes#826]
### System Info

PyTorch version: 2.5.1+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.16 (main, Dec 11 2024, 16:24:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.8.0-49-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.4.131
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA H100 80GB HBM3
GPU 1: NVIDIA H100 80GB HBM3

Nvidia driver version: 550.127.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        52 bits physical, 57 bits virtual
Byte Order:                           Little Endian
CPU(s):                               52
On-line CPU(s) list:                  0-51
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Platinum 8480+
CPU family:                           6
Model:                                143
Thread(s) per core:                   2
Core(s) per socket:                   26
Socket(s):                            1
Stepping:                             8
BogoMIPS:                             4000.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx_vnni avx512_bf16 wbnoinvd arat vnmi avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b fsrm md_clear serialize tsxldtrk avx512_fp16 arch_capabilities
Virtualization:                       VT-x
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            1.6 MiB (52 instances)
L1i cache:                            1.6 MiB (52 instances)
L2 cache:                             104 MiB (26 instances)
L3 cache:                             16 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-51
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Unknown: No mitigations
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] onnx==1.17.0
[pip3] onnxruntime==1.20.1
[pip3] torch==2.5.1
[pip3] torchvision==0.20.1
[pip3] triton==3.1.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] torch                     2.5.1                    pypi_0    pypi
[conda] torchvision               0.20.1                   pypi_0    pypi
[conda] triton                    3.1.0                    pypi_0    pypi

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### 🐛 Describe the bug

When using recipes/quickstart/inference/local_inference/multi_modal_infer.py, the generated text always includes **"end_header_id|>"** at the beginning.

**Root Cause:**

The issue arises because the **"add_special_tokens": True** parameter in **transformers/tokenization_utils_fast.py (line 526)** causes the **bos_token** to be added at the start of the prompt. This behavior is reflected in the chat template, leading to the undesired prefix in the output.

**Solution:**

To resolve this, modify line 77 in recipes/quickstart/inference/local_inference/multi_modal_infer.py as follows:

```python
inputs = processor(image, prompt, return_tensors="pt", text_kwargs={"add_special_tokens": False}).to(device)
```

### Error logs

As described above

### Expected behavior

As described above

**Comment by wukaixingxp** [comment]
@eii-lyl Thanks for report this bug. For llama model inference, we need to add special tokens like BOS. This is more of a print statement bug, because we miscalculated to length of prompt as we forgot the BOS token. I think the correct way to print it out is `processor.decode(output[0]).split( "<|start_header_id|>assistant<|end_header_id|>")[-1]` instead of `processor.decode(output[0])[len(prompt):]` in [this line](https://github.com/meta-llama/llama-recipes/blob/main/recipes/quickstart/inference/local_inference/multi_modal_infer.py#L79). Will have a quick PR to fix this.

**Comment by eii-lyl** [comment]
> @eii-lyl Thanks for report this bug. For llama model inference, we need to add special tokens like BOS. This is more of a print statement bug, because we miscalculated to length of prompt as we forgot the BOS token. I think the correct way to print it out is `processor.decode(output[0]).split( "<|start_header_id|>assistant<|end_header_id|>")[-1]` instead of `processor.decode(output[0])[len(prompt):]` in [this line](https://github.com/meta-llama/llama-recipes/blob/main/recipes/quickstart/inference/local_inference/multi_modal_infer.py#L79). Will have a quick PR to fix this.

Thank you for your reply. What you’ve described is accurate for standard LLMs. However, for VLLM models, the BOS token is already included in the chat template. When the processor adds the BOS token again, it creates the issue I’m encountering.

**Comment by wukaixingxp** [comment]
@eii-lyl Yes, you are right. I saw double BOS after `print(processor.tokenizer.decode(inputs.input_ids[0]))` 
```
<|begin_of_text|><|begin_of_text|><|start_header_id|>user<|end_header_id|>

<|image|>Describe this image<|eot_id|><|start_header_id|>assistant<|end_header_id|>
```
With your fix, now it is:
```
<|begin_of_text|><|start_header_id|>user<|end_header_id|>

<|image|>Describe this image<|eot_id|><|start_header_id|>assistant<|end_header_id|>
```
I will make a PR for it, thanks for point this out!
